pkg/controllers/dashboard/hostedcluster/controller.go

```diff
@@ -0,0 +1,128 @@
+		ReleaseNamespace:  "cattle-system",
+		ChartName:         "rancher-aks-operator-crd",
+		MinVersionSetting: settings.AKSOperatorMinVersion,
+	}
+	AksChart = chart.Definition{
+		ReleaseNamespace:  "cattle-system",
+		ChartName:         "rancher-aks-operator",
+		MinVersionSetting: settings.AKSOperatorMinVersion,
+	}
+	EksCrdChart = chart.Definition{
+		ReleaseNamespace:  "cattle-system",
+		ChartName:         "rancher-eks-operator-crd",
+		MinVersionSetting: settings.EKSOperatorMinVersion,
+	}
+	EksChart = chart.Definition{
+		ReleaseNamespace:  "cattle-system",
+		ChartName:         "rancher-eks-operator",
+		MinVersionSetting: settings.EKSOperatorMinVersion,
+	}
+	GkeCrdChart = chart.Definition{
+		ReleaseNamespace:  "cattle-system",
+		ChartName:         "rancher-gke-operator-crd",
+		MinVersionSetting: settings.GKEOperatorMinVersion,
+	}
+	GkeChart = chart.Definition{
+		ReleaseNamespace:  "cattle-system",
+		ChartName:         "rancher-gke-operator",
+		MinVersionSetting: settings.GKEOperatorMinVersion,
+	}
+)
+
+type handler struct {
+	manager *system.Manager
+	secrets v1.SecretCache
+}
+
+func Register(ctx context.Context, wContext *wrangler.Context) error {
+	h := &handler{
+		manager: wContext.SystemChartsManager,
+		secrets: wContext.Core.Secret().Cache(),
+	}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "cluster-provisioning-operator", h.onClusterChange)
+
+	return nil
+}
+
+func (h handler) onClusterChange(key string, cluster *v3.Cluster) (*v3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+
+	var toInstallCrdChart, toInstallChart *chart.Definition
+	if cluster.Spec.AKSConfig != nil {
+		toInstallCrdChart = &AksCrdChart
+		toInstallChart = &AksChart
+	} else if cluster.Spec.EKSConfig != nil {
+		toInstallCrdChart = &EksCrdChart
+		toInstallChart = &EksChart
+	} else if cluster.Spec.GKEConfig != nil {
+		toInstallCrdChart = &GkeCrdChart
+		toInstallChart = &GkeChart
+	}
+
+	if toInstallCrdChart == nil || toInstallChart == nil {
+		return cluster, nil
+	}
+
+	if err := h.manager.Ensure(toInstallCrdChart.ReleaseNamespace, toInstallCrdChart.ChartName, toInstallCrdChart.MinVersionSetting.Get(), nil, true); err != nil {
+		return cluster, err
+	}
+
+	systemGlobalRegistry := map[string]interface{}{
+		"cattle": map[string]interface{}{
+			"systemDefaultRegistry": settings.SystemDefaultRegistry.Get(),
+		},
+	}
+
+	additionalCA, err := getAdditionalCA(h.secrets)
+	if err != nil {
+		return cluster, err
+	}
+
+	chartValues := map[string]interface{}{
+		"global":               systemGlobalRegistry,
+		"httpProxy":            os.Getenv("HTTP_PROXY"),
+		"httpsProxy":           os.Getenv("HTTPS_PROXY"),
+		"noProxy":              os.Getenv("NO_PROXY"),
+		"additionalTrustedCAs": additionalCA != nil,
+	}
+
+	if err := h.manager.Ensure(toInstallChart.ReleaseNamespace, toInstallChart.ChartName, toInstallChart.MinVersionSetting.Get(), chartValues, true); err != nil {
```

Comment:
Is there a reason why ensure needs to be used twice?

Reply:
One is for CRD chart and one is for operator chart since these two are separated now.